accepted/2024/tensors/README.md

```diff
@@ -0,0 +1,424 @@
+
+`Tensor<T>` is notably explicitly sealed as extensibility must be an intentional design point. The extensibility point here is then `ITensor<TSelf, T>` and `Tensor<T>` is only our own friendly wrapper over arbitrary memory. This allows it to be more efficient and also more easily control lifetimes, ownership, and more freely extend the type in the future as needed.
+
+```csharp
+namespace System.Numerics.Tensors;
+
+public sealed class Tensor<T>
+    : IDisposable,
+      ITensor<Tensor<T>, T>
+    where T : IEquatable<T>, IEqualityOperators<T, T, bool>
+{
+    // IDisposable
+
+    void Dispose();
+
+    // ITensor
+
+    public static Tensor<T> Empty { get; }
+
+    public bool IsEmpty { get; }
+    public bool IsPinned { get; }
+    public int Rank { get; }
+    public ReadOnlySpan<nint> Strides { get; }
+
+    public ref T this[params ReadOnlySpan<nint> indices] { get; }
+
+    public static implicit operator TensorSpan<T>(Tensor<T> value);
+    public static implicit operator TensorReadOnlySpan<T>(Tensor<T> value);
+
+    public TensorSpan<T> AsSpan(params ReadOnlySpan<Range<nint>> ranges);
+    public ref T GetPinnableReference();
+    public Tensor<T> Slice(params ReadOnlySpan<Range<nint>> ranges);
+
+    public void Clear<T>();
+    public void CopyTo<T>(TensorSpan<T> destination);
+    public void Fill<T>(T value);
+    public bool TryCopyTo<T>(TensorSpan<T> destination);
+
+    // IEnumerable
+
+    public Enumerator GetEnumerator();
+
+    // IEquatable
+
+    public bool Equals(Tensor<T> other);
+
+    // IEqualityOperators
+
+    public static bool operator ==(Tensor<T> left, Tensor<T> right);
+    public static bool operator ==(Tensor<T> left, T right);
+
+    public static bool operator !=(Tensor<T> left, Tensor<T> right);
+    public static bool operator !=(Tensor<T> left, T right);
+
+    public struct Enumerator
+    {
+        public ref readonly T Current { get; }
+
+        public bool MoveNext();
+    }
+}
+
+public static partial class Tensor
+{
+    // Effectively mirror the TensorPrimitives surface area. Following the general pattern
+    // where we will return a new tensor and an overload that takes in the destination explicitly.
+    // * public static ITensor<T> Abs<T>(ITensor<T> x) where T : INumberBase<T>;
+    // * public static void Abs<T>(ReadOnlyTensorSpan<T> x, TensorSpan<T> destination) where T : INumberBase<T>;
+
+    // The language ideally has extension operators such that we can define `operator +` instead of `Add`
+    // Without this support, we end up having to have `TensorNumber<T>`, `TensorBinaryInteger<T>`, etc
+    // as we otherwise cannot correctly expose the operators based on what `T` supports
+
+    // APIs that would return `bool` like `GreaterThan` are split into 3. Following the general
+    // pattern already established for our SIMD vector types.
+    // * public static ITensor<T> GreaterThan(ITensor<T> x, ITensor<T> y);
```

Comment:
This returns `ITensor<T>` rather than `ITensor<bool>`? I understand why we do `Vector<T>` for the vector types, but that doesn't seem like the right answer here.

Reply:
Typo. This is meant to be `ITensor<bool>` here.

Notably this is one of the places where what's best depends on what `T` is. If its a primitive type and going to be vectorized, `bool` is a terrible option. If its a user-defined type, then its one of the better choices. Some people may even want a bitmask result instead.

I want us to discuss this a little bit and see if we also want to expose any special APIs for the case of the user wanting `ITensor<T>` or `BitMask` returned or if we want to hold off on that until there's a need. -- Functionally we'll end up implementing the former and then narrowing to `byte` before storage for the vectorized case regardless, so exposing it is just a question of naming the API we'll already be writing.